runtime/lua/vim/diagnostic.lua

```diff
@@ -346,18 +366,15 @@ local function schedule_display(namespace, bufnr, args)
   bufs_waiting_to_update[bufnr][namespace] = args
 
   local key = make_augroup_key(namespace, bufnr)
+  local group = vim.api.nvim_create_augroup(key, { clear = true })
   if not registered_autocmds[key] then
-    vim.cmd(string.format(
-      [[augroup %s
-      au!
-      autocmd %s <buffer=%s> lua vim.diagnostic._execute_scheduled_display(%s, %s)
-    augroup END]],
-      key,
-      table.concat(insert_leave_auto_cmds, ','),
-      bufnr,
-      namespace,
-      bufnr
-    ))
+    vim.api.nvim_create_autocmd(insert_leave_auto_cmds, {
+      group = group,
+      buffer = bufnr,
+      callback = function()
+        execute_scheduled_display(namespace, bufnr)
+      end,
+    })
     registered_autocmds[key] = true
```

Comment:
`registered_autocmds` could now use group IDs instead of names, and I think a similar thing could be done with `close_preview_autocmd` in `vim.lsp.util`, but maybe that's for future PRs.